src/functions_topic.py

```diff
@@ -0,0 +1,45 @@
+import tester
+
+
+def concatenate_strings(string1: str, string2: str) -> str:
+    """
+    Returns the concatenation of two strings.
+    E.g.
+        concatenate_strings("hello", "world") -> "helloworld"
+
+    @rtype: str
+    @param string1: First string of the concatenation
+    @param string2: Second string of the concatenation
+    @return: The concatenation of string1 and string2
+    """
+    concatenated = string1 + string2
+    return concatenated
+
+
+# Tests for the concatenate_strings method
+tester.test_function("", concatenate_strings, "", "")
+tester.test_function("ab", concatenate_strings, "a", "b")
```

Comment:
This is... not normal. Using a higher order function to test is pretty rough for potentially people with literally no coding experience.

Reply:
Any thoughts on an alternative? I figured this would be easiest for them, but there is defo a lot of "magic code" in here.

Reply:
I suppose just writing out the assert on their own would work. 

`assert "ab" == vincat("a", "b")`

Reply:
`assert`s don't do anything in normal execution typically -- I'm not sure what python's behaviour is there, but I imagine it'd be the same as many other languages.

Reply:
I think it throws an assertion error, but I'll look into this when I'm supposed to be lecturing.

Reply:
It likely only does so when debugging flags are enabled.

Reply:
> Any thoughts on an alternative?

Do you touch on classes and inheritance? Maybe testing can be introduced after that topic and you use the built-in `unittest`.

Reply:
Inheritance isn't covered, and testing is one of the last topics. 

Additionally, it appears to be a nightmare getting `unittest` to work on colab and requires mounting your Google drive (that said, a quick googz seems to suggest otherwise --- maybe this was changed recently-ish?). I'll do a little more digging here. 

Sadly, still, it would have to be hit late in the course and it would be good to cover testing earlier.

Reply:
Looks like colab will throw the error regardless. 

![image](https://user-images.githubusercontent.com/1328048/157549398-cc5c3eca-667c-492a-932e-e6e5a15d94d0.png)

Reply:
Nice. Either they run in debug mode or python just always behaves that way. My guess is the former, without looking into it at all.

Reply:
[Looks like `unittest` is actually easy in colab.](https://colab.research.google.com/github/damorimRG/practical_testing_book/blob/master/testgranularity/unittesting.ipynb)

That said, it requires classes which is a late course topic. Solution --- Use simple asserts to start. When objects are hit, we can use `unittest`.

Reply:
That's probably reasonable enough. Think it makes sense just to get the idea across of semi automated testing, then be like "yo dawg this is actually your life" much later.